docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+      permissions: '0644'
+    runcmd:
+      - /usr/bin/machineadm --bootstrap --path /tmp/machineadm_config.yaml
+```
+#### MachineConfig types
+
+It is important to note that we have in effect two representations of the MachineConfig: First, we have the
+MachineBootstrap Kubernetes resource that may reference other Kubernetes resources such as secrets for inclusion
+in the machine configuration. Secondly, we have the "serialized" MachineConfiguration which is the "fully realized"
+configuration, that inlines all of the referenced data and is stored as its own Kubernetes secret for handoff to
+the infrastructure provider for secure storage.
+
+##### Common types
+
+```go
+// MachineBootstrapConfig is a resource that defines the configuration
+// of a machine
+type MachineBootstrapConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineBootstrapConfigSpec   `json:"spec,omitempty"`
+	Status MachineBootstrapConfigStatus `json:"status,omitempty"`
+}
+
+// MachineBootstrapConfigSpec defines the desired state of a MachineBootstrapConfig
+type MachineBootstrapConfigSpec struct {
+	PluginRefs []corev1.ObjectReference `json:"pluginRefs"`
+}
+
+// MachineBootstrapConfigTemplate is a resource that defines the configuration
+// of a machine
+type MachineBootstrapConfigTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec MachineBootstrapConfigTemplateSpec `json:"spec,omitempty"`
+}
+
+// MachineBootstrapConfigTemplateSpec defines the desired state of a MachineBootstrapConfigTemplate
+type MachineBootstrapConfigTemplateSpec struct {
+	MachineBootstrapConfigSpec `json:",inline"`
+	// PluginTemplateRefs is a list of machine bootstrap plugin template refs
+	// which will be instantiated for each machine.
+	PluginTemplateRefs []corev1.ObjectReference `json:"pluginTemplateRefs,omitempty"`
+}
+
+// MachineBootstrapConfigStatus defines the observed state of KubeadmConfig
+type MachineBootstrapConfigStatus struct {
+	// Ready indicates the BootstrapData field is ready to be consumed
+	Ready bool `json:"ready,omitempty"`
+
+	// DataSecretName is the name of the secret that stores the bootstrap data script.
+	// +optional
+	DataSecretName *string `json:"dataSecretName,omitempty"`
+
+	// FailureReason will be set on non-retryable errors
+	// +optional
+	FailureReason string `json:"failureReason,omitempty"`
+
+	// FailureMessage will be set on non-retryable errors
+	// +optional
+	FailureMessage string `json:"failureMessage,omitempty"`
+
+	// ObservedGeneration is the latest generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Conditions defines current service state of the KubeadmConfig.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// PluginStatus defines the status of a machineadm plugin
+type PluginStatus struct {
```

Comment:
What is the expected usage of this type?

Reply:
Bootstrap reporting